tephra-core/src/main/java/org/apache/tephra/persist/AbstractTransactionLog.java

```diff
@@ -48,21 +52,30 @@
   protected long timestamp;
   private volatile boolean initialized;
   private volatile boolean closed;
+  private AtomicLong writtenUpTo = new AtomicLong();
   private AtomicLong syncedUpTo = new AtomicLong();
```

Comment:
And this can be `long`.

Reply:
Actually not. Just notice it is used outside of writer sync block.

Reply:
but, please make it `final`.

Reply:
or a `volatile long` also work since we only use `get` and `set`.

Reply:
good suggestion, done